Fix warped and crimson fungi not being able to be placed on nylium.

Fealtous opened the pull request:
Fix for #10543. I will see about a game test for this in a bit.

Fealtous committed:
Fix warped and crimson fungi not being able to be placed on nylium.

Jonathing commented:
PlantType 🙃

Fealtous commented:
> PlantType 🙃

thaaat's promising...

Fealtous committed:
Fix other PlantType parity issues, add test for all vanilla PlantType blocks. Will probably need to write another test for mod stuff.

Fealtous committed:
Fix placing mushrooms on glass. Game test is failing when setblocking some stone, very confused.

hunterw523 commented:
That is a wonderful change. I've actually seen a player get really frustrated by this. How exactly did you fix this?

Fealtous commented:
> That is a wonderful change. I've actually seen a player get really frustrated by this. How exactly did you fix this?

Short answer is I updated our patch in Block and added the proper entries into PlantType and IPlantable.

The longer answer is that this hook is really old and needed some love. The original reported bug was for fungus, but there were actually several parity issues that had snuck in from the latest few mc versions (or when updating to 1.21.5, unclear when they popped up), being:

Mushrooms could be placed on glass and other non-solid blocks.
Mangrove propagules and the azaleas couldn't be planted on clay or farmland.
Deadbushes couldn't be planted on all colors of terracotta or the various dirt types when they should, but could be planted on all glazed terracotta when they shouldn't.

So yeah, a number of issues, though let's be real-- no one was going to notice the dead bush one.
The current task is creating the test mod to validate vanilla behavior, which is technically done, but needs to be split for in-game size reasons. This section of forge isn't pretty, but having a thorough test mod should keep us in check for future updates.

Fealtous committed:
dynamic test gen?

LexManos committed:
Better tests

LexManos committed:
Add TODO test to fix coverage test.
I'm happy with how much is covered and can add the rest later.

LexManos commented:
https://github.com/MinecraftForge/MinecraftForge/compare/4afa7384ed87fcdf49c0bb609def26824a74dd63...LexManos:MinecraftForge:plant_type_tests

Fixed the tests, there are still some vanilla blocks not covered. But covers a lot more then yours does.
Pull that in to your PR and poke me to pull it.

Fealtous commented:
![image](https://github.com/user-attachments/assets/007ec755-8688-4479-bf51-72d049812045)
